574 - Support for Dark mode

tinahollygb opened the pull request:
### Features and Changes

Adds support for dark mode that:

- uses the users default operating system setting (if available)
- allows the user to override their operating system setting

I've gone through all of the screens that I could find. For ones I couldn't find, I made a best guess based on the colours I saw. 

Changes:

- Bootstrap components, as well as CSS-in-JS components needed to be overridden. 


Things I didn't change:

- most things with purple. So buttons remain as they were. Also the sidebar remained unchanged as it is dark mode-friendly enough.


Issues:

- Closes #574 


### How to be Dark-mode friendly as a developer

I will write a proper document on Notion but for now, the TL;DR is:

1. use the CSS variables, i.e. `var(--text-color-main)` instead of `$black` so that it will flip properly
2. when adding a new library that uses CSS, test it in dark mode and ensure it looks good. if it doesn't have the right theme, override the colours as I did (see the `bootstrap-theme-overrides` file).
3. if you need to add a new colour, add it to both the light and dark mixins in the `theme.scss` file.
4. test and when putting up a PR, include screenshots of both light and dark modes


### Dependencies

n/a

### Testing

Test all the things.

### Screenshots

#### Latest Icons

Latest icons (not reflected in below screenshots).

![image](https://user-images.githubusercontent.com/113377031/193951177-8de8fa96-0daf-4c62-9347-a83283fdab88.png)
![image](https://user-images.githubusercontent.com/113377031/193951189-553f89bb-86a9-4afe-bd9e-1b94103ca000.png)
![image](https://user-images.githubusercontent.com/113377031/193951199-51ab963a-8f9d-4cee-ae65-a2db53f513ff.png)


#### Top Bar

![CleanShot 2022-09-30 at 12 38 43@2x](https://user-images.githubusercontent.com/113377031/193344344-b9e967f7-3fa6-4418-bae1-af7e769e1b60.png)

<img width="1076" alt="image" src="https://user-images.githubusercontent.com/113377031/193344747-707e0d50-2e82-4826-9ffc-f94fe3fed8f9.png">

<img width="1074" alt="image" src="https://user-images.githubusercontent.com/113377031/193353038-ad81c4a7-6a15-476c-9b3f-195805ed8c21.png">


#### Sidebar

No changes to the theme, just the dropdowns to be dark-friendly.

<img width="255" alt="image" src="https://user-images.githubusercontent.com/113377031/193344927-b190725f-2879-4c13-b3ff-3fc764f4cdd3.png">
<img width="249" alt="image" src="https://user-images.githubusercontent.com/113377031/193344970-c0f2466a-f929-44d8-9a18-0d8b28341b86.png">

#### Forms and Alerts

Forms were styled to be dark-friendly.

![CleanShot 2022-09-30 at 12 45 14@2x](https://user-images.githubusercontent.com/113377031/193345324-e32b64cc-d65d-44f3-a941-013ebd81300b.png)

Alerts too. 
<img width="1255" alt="Screen Shot 2022-09-29 at 6 30 18 PM" src="https://user-images.githubusercontent.com/113377031/193345349-0d078908-c8bc-4796-ac57-bfaee22019f1.png">
<img width="1245" alt="Screen Shot 2022-09-29 at 6 32 13 PM" src="https://user-images.githubusercontent.com/113377031/193345354-de1e429f-6719-4ec1-8cc9-7ae8caf669c1.png">
<img width="1247" alt="Screen Shot 2022-09-29 at 6 36 50 PM" src="https://user-images.githubusercontent.com/113377031/193345357-2db1a07d-4793-493d-b2e8-7565dad65f6b.png">

#### Features

<img width="1233" alt="image" src="https://user-images.githubusercontent.com/113377031/193346043-5389b522-2e83-4bfa-a9d2-756bb9dfc29c.png">
<img width="1224" alt="image" src="https://user-images.githubusercontent.com/113377031/193346114-6ab3abf6-ed1e-4e7c-9dfd-0574580b34af.png">

#### Experiments

<img width="1224" alt="image" src="https://user-images.githubusercontent.com/113377031/193346238-ef5f2b90-1e9d-4d4d-a87a-a5b44401bd98.png">

We can maybe improve the placeholder images by using transparent images.

<img width="1224" alt="image" src="https://user-images.githubusercontent.com/113377031/193346333-7c76bc07-3d2a-4299-99bd-3c1010135686.png">

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/113377031/193346387-5f0792fa-a78d-4a42-adb6-d4b3d5ee0f39.png">

Diff viewer.

<img width="1009" alt="image" src="https://user-images.githubusercontent.com/113377031/193352724-c0832e5d-e2b8-475d-a555-c0f93b8f4c13.png">
<img width="999" alt="image" src="https://user-images.githubusercontent.com/113377031/193352794-7364bb79-3d5c-400a-be70-7dbb99cbc835.png">


#### Metrics

<img width="1239" alt="image" src="https://user-images.githubusercontent.com/113377031/193346444-76574d3d-0f66-4f83-89c9-5acd5063ab3c.png">
<img width="1241" alt="image" src="https://user-images.githubusercontent.com/113377031/193346486-e0c7ef5a-f0b9-4295-8094-8d396f001c3e.png">

#### Segments

<img width="1234" alt="image" src="https://user-images.githubusercontent.com/113377031/193346539-116a9f46-b52f-40db-9cad-4459c2088c62.png">

#### Dimensions

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/113377031/193346925-f797f49c-21e5-455a-aecb-3a658b8ca490.png">

#### Data Sources

<img width="1248" alt="image" src="https://user-images.githubusercontent.com/113377031/193347157-5aeeeb80-c584-44e9-bdc8-487ed0083143.png">
<img width="1239" alt="image" src="https://user-images.githubusercontent.com/113377031/193347187-f1f99f23-ffd8-4331-a624-7b722eeddb04.png">
<img width="1033" alt="image" src="https://user-images.githubusercontent.com/113377031/193349947-0be8e320-ad3e-41b2-a67f-50027d6d0667.png">

For the Code component that uses AceEditor, I opted for theme `tomorrow_night` instead of `textmate` for dark mode.

<img width="1236" alt="image" src="https://user-images.githubusercontent.com/113377031/193351887-6ea8fd4b-1dc8-462b-97d6-c3d4d6322e4e.png">

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/113377031/193352150-ebec888a-a82f-4937-80b4-2e299b94abc1.png">

<img width="1224" alt="image" src="https://user-images.githubusercontent.com/113377031/193352870-a5b596b5-00db-42e3-9053-6672870ded2c.png">


#### Import config.yml differ

<img width="1221" alt="image" src="https://user-images.githubusercontent.com/113377031/193349856-e6772acb-4ca9-486e-9392-c3f9fc261d4f.png">

#### Management

<img width="1009" alt="image" src="https://user-images.githubusercontent.com/113377031/193352289-8efcd9b7-ae51-4b87-a92f-2a088aef9094.png">

<img width="376" alt="image" src="https://user-images.githubusercontent.com/113377031/193352337-1d8cc8a2-d2fe-4491-8298-43334a8496df.png">
<img width="1023" alt="image" src="https://user-images.githubusercontent.com/113377031/193352363-0d7fee7c-1b0d-406e-939b-6d608495271f.png">
<img width="1222" alt="image" src="https://user-images.githubusercontent.com/113377031/193352392-5e672109-2489-4265-9cc6-20192edf6283.png">
<img width="1206" alt="image" src="https://user-images.githubusercontent.com/113377031/193352430-8c27791d-2402-4162-8a71-6c0664472b41.png">

#### Settings

<img width="1014" alt="image" src="https://user-images.githubusercontent.com/113377031/193352505-9f7a0268-a861-4e4a-b66a-7e3bcabe9a49.png">
<img width="1018" alt="image" src="https://user-images.githubusercontent.com/113377031/193352536-5c02787e-a516-4b28-b8d8-b6d18631efa9.png">

#### Visualizations

![image](https://user-images.githubusercontent.com/113377031/193696513-4cd92931-581e-4070-960b-c6bf70c5c6cd.png)
<img width="710" alt="image" src="https://user-images.githubusercontent.com/113377031/193696592-f502e21d-aed7-4dd9-8161-ba103ed1d92e.png">

![image](https://user-images.githubusercontent.com/113377031/193696358-5460b884-de1b-4520-8ea2-d413c5b29384.png)
![image](https://user-images.githubusercontent.com/113377031/193696363-d659e3b5-053d-447b-9103-f3cbc0752d57.png)
![image](https://user-images.githubusercontent.com/113377031/193696380-eeb6995e-4eb4-4f66-ab77-5f82e0470130.png)
![image](https://user-images.githubusercontent.com/113377031/193696393-5f1cbb2e-8b15-49d8-a0cc-5d4a370d55b4.png)



<img width="671" alt="image" src="https://user-images.githubusercontent.com/113377031/193681677-def058f4-cbbc-4aab-966d-e480271ba8dc.png">
<img width="1077" alt="image" src="https://user-images.githubusercontent.com/113377031/193682413-3480c7f4-82c8-4963-9ef0-a712e4791b74.png">

Draw
<img width="269" alt="image" src="https://user-images.githubusercontent.com/113377031/193685149-71edaa68-a02d-462d-8af1-35a0315cda83.png">

Warning
<img width="715" alt="image" src="https://user-images.githubusercontent.com/113377031/193685227-421f54a3-3def-41a2-a317-9d52df052b81.png">





#### Getting Started flow

Expand the toggle section below to see all the screenshots.

<details>
<summary><b>Getting Started section</b></summary>

<img width="1151" alt="image" src="https://user-images.githubusercontent.com/113377031/193345703-5d96cb24-f8bb-41e9-ba3d-eb3cfa9a074b.png">
<img width="1184" alt="image" src="https://user-images.githubusercontent.com/113377031/193345738-247e3796-7362-4a10-ad21-c25ad61f5747.png">
<img width="1065" alt="image" src="https://user-images.githubusercontent.com/113377031/193345752-7fbb19d6-10d2-4222-a0f2-4e9c378529d1.png">
<img width="1076" alt="image" src="https://user-images.githubusercontent.com/113377031/193345769-207bfe05-6c74-4f8c-b8e8-daa4692d5d3d.png">
<img width="1067" alt="image" src="https://user-images.githubusercontent.com/113377031/193345785-bbe1663d-88de-4684-bb9d-4c25c9544aab.png">
<img width="1073" alt="image" src="https://user-images.githubusercontent.com/113377031/193345804-89b42b51-7694-4da2-a94f-154e73a198e1.png">

<img width="1055" alt="image" src="https://user-images.githubusercontent.com/113377031/193345869-f7aaddd8-72be-4a18-896c-eb06f7413ca8.png">
<img width="1188" alt="image" src="https://user-images.githubusercontent.com/113377031/193345903-82944dcd-a790-4d78-b086-e3fb409f9501.png">
<img width="1053" alt="image" src="https://user-images.githubusercontent.com/113377031/193345921-21fd2e20-58e7-4802-b3d2-d2c550cdfb3c.png">


<img width="1053" alt="image" src="https://user-images.githubusercontent.com/113377031/193345848-27c1dd9a-c62d-444d-aa71-fb34b24b6e6f.png">


</details>

tinahollygb commented on packages/front-end/components/Forms/CodeTextArea.tsx:
```diff
@@ -49,6 +53,28 @@ export default function CodeTextArea({
     otherProps.error = semicolonWarning;
   }
 
+  const theme = useMemo(() => {
+    let actualTheme = LIGHT_THEME;
+
+    try {
+      const fromStorage = localStorage.getItem("gb_ui_theme");
+      if (
+        !fromStorage &&
+        window.matchMedia("(prefers-color-scheme: dark)")?.matches
+      ) {
+        actualTheme = DARK_THEME;
+      }
+
+      if (fromStorage === "dark") {
+        actualTheme = DARK_THEME;
+      }
+    } catch (e) {
+      return actualTheme;
+    }
+
+    return actualTheme;
+  }, []);
```
When we need to do something like this again, it would be a good time to do a refactor to add it as a context provider and expose a `useAppearanceTheme()` hook or similar.

tinahollygb commented on packages/front-end/components/Layout/ThemeToggler.tsx:
```diff
@@ -0,0 +1,112 @@
+
+  useGlobalMenu(".top-nav-theme-menu", () => setThemeDropdownOpen(false));
+
+  useEffect(() => {
+    try {
+      const themeFromStorage = localStorage.getItem("gb_ui_theme") as
+        | UIAppearanceMode
+        | undefined;
+
+      if (themeFromStorage === "light" || themeFromStorage === "dark") {
+        document.documentElement.className = `theme--${themeFromStorage}`;
+        setActiveTheme(themeFromStorage);
+      }
+    } catch (e) {
+      // We are unable to retrieve the theme changes due to the browser's privacy settings
+    }
+  }, []);
+
+  const handleThemeChange = useCallback((theme: UIAppearanceMode) => {
+    ["theme--dark", "theme--light"].forEach((c) => {
+      document.documentElement.classList.remove(c);
+    });
+
+    setActiveTheme(theme);
+
+    setThemeDropdownOpen(false);
+
+    try {
+      if (theme === "system") {
+        localStorage.removeItem("gb_ui_theme");
+      } else {
+        document.documentElement.classList.add(`theme--${theme}`);
+        localStorage.setItem("gb_ui_theme", theme);
+      }
+    } catch (e) {
+      // We are unable to persist the theme changes due to the browser's privacy settings
+    }
+  }, []);
```
Related to the last comment, we would refactor this into a context provider and this component would just be UI that links up the functions from that provider.

tinahollygb commented on packages/front-end/styles/_bootstrap-theme-overrides.scss:
```diff
@@ -0,0 +1,372 @@
+// Global Bootstrap overrides that respect dark/light modes
```
These are mostly Bootstrap but we have other components that are overridden here.

tinahollygb commented on packages/front-end/styles/_bootstrap-theme-overrides.scss:
```diff
@@ -0,0 +1,372 @@
+
+.form-group,
+.gb-select,
+.gb-multi-select {
+  [class$="-menu"] {
+    background-color: var(--surface-background-color);
+  }
+
+  [class$="-option"] {
+    color: var(--text-color-main);
+  }
+
+  [class$="-Input"] {
+    color: var(--text-color-main);
+  }
+
+  [class$="-singleValue"] {
+    color: var(--text-color-muted);
+  }
+
+  [class$="-control"] {
+    background-color: var(--surface-background-color);
+    border-color: var(--border-color-300);
+  }
+
+  [class$="-multiValue"] {
+    background-color: var(--form-multivalue-background-color);
+
+    > * {
+      color: var(--form-multivalue-text-color);
+    }
+  }
+}
```
These selectors are quite fragile. If we update the library we use for multiple select, there's a good chance this will break. That said, if we do that, it's likely the developer would test that and we'd catch it, so probably ok to leave them.

tinahollygb commented on packages/front-end/styles/_bootstrap-theme-overrides.scss:
```diff
@@ -0,0 +1,372 @@
+
+// region d2h differ
+
+// d2h loads CSS at the end of the document so we need to do all these !important
```
This is the differ used in the import config.yml section.

tinahollygb commented on packages/front-end/styles/_bootstrap-theme-overrides.scss:
```diff
@@ -0,0 +1,372 @@
+
+// region Diff Viewer
+
+.diff-wrapper {
```
This is the differ used in the history panel.

tinahollygb commented on packages/front-end/styles/_bootstrap-variable-overrides.scss:
```diff
@@ -0,0 +1,13 @@
+$font-size-base: 0.875rem; // change the default from 16px to 14px.
```
Original overrides moved to a separate file.

jdorn commented on packages/front-end/components/Layout/ThemeToggler.tsx:
```diff
@@ -0,0 +1,112 @@
+        }}
+        className="btn mr-1 text-secondary"
+      >
+        <FaMoon />
```
Should we change the icon here depending on what theme they have selected?  Also, I think we should try to give some indication that this is clickable and what it does.  Could be as simple as adding a dropdown arrow.  Or a title/tooltip saying "Change theme".

tinahollygb replied:
I've updated the toggler to change the icon based on the preferred theme: system, dark, light.

jdorn commented on packages/front-end/components/Forms/CodeTextArea.tsx:
```diff
@@ -49,6 +53,28 @@ export default function CodeTextArea({
+    }
+
+    return actualTheme;
+  }, []);
```
We need this same logic in the `Code` component.  It already supports light/dark mode, but the theme is passed in as a prop right now instead of using the actual selected theme.

It's currently using the "light" theme to render JSON on the feature page and the SQL/Python code for data sources:

![image](https://user-images.githubusercontent.com/1087514/193817652-445a0fc4-45ea-41b5-9042-10b4f81cace8.png)
![image](https://user-images.githubusercontent.com/1087514/193818204-40c90d12-97c9-426a-a17d-67e6080c174b.png)


And it's using the "dark" theme when rendering code samples and SQL queries on the metric page:

![image](https://user-images.githubusercontent.com/1087514/193818464-a22265be-1aac-4b70-b8a5-f56fdba5c967.png)

![image](https://user-images.githubusercontent.com/1087514/193818580-3eee4c96-efe9-4ee9-8dad-5680e6e8a902.png)

tinahollygb replied:
I've updated all code blocks to use the theme from the context provider.

![CleanShot 2022-10-04 at 14 08 53@2x](https://user-images.githubusercontent.com/113377031/193929344-d2b41898-6cf7-4564-924a-1643d908feff.png)
![CleanShot 2022-10-04 at 14 09 50@2x](https://user-images.githubusercontent.com/113377031/193929570-b3f7d1c9-5ff4-42e7-af37-070d5b04fc9d.png)

<img width="399" alt="image" src="https://user-images.githubusercontent.com/113377031/193929715-7ff34ffd-f063-4ce0-8c8a-812284909ba0.png">
<img width="347" alt="image" src="https://user-images.githubusercontent.com/113377031/193929758-9c72c853-89fa-4944-85a7-3a2c72a46bb3.png">

jdorn reviewed (commented):
Found a few edge cases in dark mode:

### Upgrade modal on Cloud
![Screenshot from 2022-10-04 07-35-17](https://user-images.githubusercontent.com/1087514/193821745-a068004c-85a4-4413-b590-f44f0629ccb2.png)

To test this, you'll need to render it manually somewhere:

```tsx
      <UpgradeModal
        close={() => {console.log("close")}}
        source="top-nav-freeseat-overage"
        reason="Whoops! You are over your free seat limit."
      />
```

### Activity Feed
Click the Bell icon in the top nav
![Screenshot from 2022-10-04 07-31-46](https://user-images.githubusercontent.com/1087514/193822141-296c8b91-bdb4-44a1-9d93-0b1fc2d73edb.png)

### Review Feature Changes
When you make changes to a feature and review them before publishing
![Screenshot from 2022-10-04 07-30-11](https://user-images.githubusercontent.com/1087514/193822317-d14d7280-2b07-4a6b-b013-0534e0ee2c45.png)

### Namespace Settings
First, add a namespace in Settings.  Then, add an experiment rule to a feature and enable namespaces.  The legend and purple number are really hard to read in dark mode.
![image](https://user-images.githubusercontent.com/1087514/193822652-c643e14d-90ba-4ba6-8f2a-a06725a5fc7a.png)

tinahollygb commented:
Updated the loading spinner in the modal.

<img width="568" alt="image" src="https://user-images.githubusercontent.com/113377031/193898426-24d442c9-d03c-4a04-bc3f-05b74dcb2f38.png">

I've also updated the dropdown icon to reflect the currently-selected theme:

<img width="288" alt="image" src="https://user-images.githubusercontent.com/113377031/193928277-7f5e2832-4414-4734-812b-0c282c2eb13e.png">
<img width="345" alt="image" src="https://user-images.githubusercontent.com/113377031/193928313-d28d23e0-48e0-49b0-959d-767b14d4a236.png">
<img width="273" alt="image" src="https://user-images.githubusercontent.com/113377031/193928341-6adc75ec-41da-4e02-9628-cde136a3d3ea.png">

To support the multiple areas where we need to get the current theme, I refactored the theme logic into a context provider.

tinahollygb commented on packages/front-end/styles/theme.scss:
```diff
@@ -0,0 +1,126 @@
+
+  @media (prefers-color-scheme: dark) {
+    @include dark-theme();
+  }
```
`prefers-color-scheme` has been [available since 2019](https://caniuse.com/prefers-color-scheme) in Chrome, Edge, Firefox, Safari and Opera so I think it should be ok, but I've defaulted to light just in case.

tinahollygb commented:
I've updated the code fold section:

<img width="1362" alt="image" src="https://user-images.githubusercontent.com/113377031/193933015-38531c93-130a-4043-b6e6-b46e45ec8860.png">

The review feature changes diff has been updated. It uses the `ExpandableDiff` component. So far I've found 3 different components for diffs. The project implements 2 diff plugins so far. Do you know if we have any more instances of diffs than the ones I've already found? If they implement the same one this section uses, it'll need to be wrapped in `diff-wrapper` to have the changes applied.

<img width="843" alt="image" src="https://user-images.githubusercontent.com/113377031/193934045-4db1a783-e40c-4fa4-8f97-125b8f2c4954.png">

tinahollygb commented:
I have updated the namespace background too.

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/113377031/193934577-c435e9c3-2901-4e1b-8e86-3090462be9d1.png">

tinahollygb commented:
I've also improved the placeholder image appearance.

<img width="667" alt="image" src="https://user-images.githubusercontent.com/113377031/193935455-3d266657-79c7-4019-8d90-9a1f8310e9a4.png">
<img width="690" alt="image" src="https://user-images.githubusercontent.com/113377031/193935497-76f961b3-ab95-4e21-b7f2-f0c2273693ed.png">

This was the original, for reference. The following old images are now outdated:

![CleanShot 2022-10-04 at 14 41 38@2x](https://user-images.githubusercontent.com/113377031/193935712-23b7b2cf-570c-4b63-baf7-9ce1f772aa16.png)
![CleanShot 2022-10-04 at 14 42 06@2x](https://user-images.githubusercontent.com/113377031/193935756-f11252db-c63c-4edc-b254-8c692eda354c.png)

While the contrast is poor in dark mode, I think it's preferred over the bright image, so I've made the change. Let me know if you'd like me to revert back to the previous placeholder style.

tinahollygb commented:
@jdorn I believe I have implemented the change requests. You can see individual comments with screenshots but here is a summary:

- fixed the dark mode modal loading state overlay
- fixed the code folding area in the activity feed diff
- added the `diff-wrapper` class to the `ExpandableDiff` component so it could use the defined styles 
- improved background and border colours for the namespace visualization
- defaulted to light theme should `prefers-color-scheme` not be available
- implemented the dynamic theme in the `<Code />` block, which updates all instances to use either light or dark depending on the theme chosen. This will have the change that in light mode it will use the light theme, which will be different than previous behaviour.
- refactored the theme logic into a context provider to expose a hook so we can get the current theme (light, dark) as well as the user-preferred theme (system, light, dark).
- adjusted the placeholder image background—this causes it to have low contrast in dark mode but it's less jarring

Let me know if I missed anything or if there's anything else that needs to be changed.
